Fix EditorTitleBar excessive width

WhalesState opened the pull request:
https://github.com/user-attachments/assets/8113c691-19dc-42b6-aae9-5de3d632c9c4

Summary of Changes:
* **EditorScrollBox Extracted for Reuse**:
* * The EditorScrollBox (introduced in https://github.com/godotengine/godot/pull/97878) has been extracted into its own class to allow reuse across the editor. This helps address excessive width issues.

* **Improved Layout in EditorTitleBar**:
* * Refactored the layout using three expanding containers instead of relying on spacer size calculations. This simplifies alignment and makes future layout changes easier:
`[ left_hb                 ] [             center_hb             ] [                 right_hb ]`

* **New Minimum Editor Size**:
* * The minimum size of the editor window is now set to 640×480. This change required adjusting many UI elements to fit within the new constraints. (Most of those adjustments are already done locally).
 
* **Selective Use of EditorScrollBox**:
* * The new EditorScrollBox has been applied to the main menus and screen buttons, but not to right_hb, which includes the EditorRunBar and plugin-specific controls.
Note: it’s currently difficult to keep the main screen buttons centered when many plugins are added to the right container.
 
Requires testing on Android Editor to confirm compatibility and layout integrity on smaller screens.

* Note: The change to the Renderer Button shown in the demo video is not included in this PR. That change is part of https://github.com/godotengine/godot/pull/109357, and is shown in the video only to demonstrate the final result when both PRs are merged.

* Note: The Pause and Stop button changes shown in the video have been removed from this PR. They will be moved to a separate PR if needed.

KoBeWi commented:
For the main screen buttons, I was wondering if we shouldn't change them to TabBar/TabContainer, which would eventually allow to use main screen as a dock slot. Related: #106164

arkology commented:
For main screen buttons this is 100% improvement (existing buttons will not break layout at editor minimum size and buttons added by plugins will not make it worse).
Other changes I think require more discussion.
In my opinion the end goal of all this changes (including https://github.com/godotengine/godot/pull/102301) is to fit editor on a quarter (at least) of the screen (fullhd monitor).

WhalesState commented:
> For the main screen buttons, I was wondering if we shouldn't change them to TabBar/TabContainer, which would eventually allow to use main screen as a dock slot. Related: #106164

I already have an old change that allows reparenting the `TabBar` of the `TabContainer` which may help with this, The only issue is that all the Main Screen plugins have a very large width due to their top bar minimum width which will make it harder to align to of them horizontally.

<img width="615" height="45" alt="image" src="https://github.com/user-attachments/assets/d622f54e-133f-4637-be8f-5b02a94921e9" />


I already have some changes locally that can reduce their width and was thinking about reusing the new EditorScrollBar container introduced in this PR to move the select buttons to the left of the 2D/3D Editors like how they are done in Blender, I remember i have seen a feature request asking for that too.

<img width="264" height="440" alt="image" src="https://github.com/user-attachments/assets/dc80a12c-62d2-456f-afcc-ca6f6922c08a" />

WhalesState commented:
> For main screen buttons (2D, 3D, ect.) this is 100% improvement (existing buttons will not break layout at editor minimum size and buttons added by plugins will not make it worse). Other changes I think require more discussion. In my opinion the end goal of all this changes (including #102301) is to fit editor on a quarter (at least) of the screen (fullhd monitor). For android - being able to use editor in portrait mode.

* Yes the main goal is to support smaller screens for the editor to be able to increase the editor scale or tile the editor window without issues, and the only issue we have now are bottom panels and the main screen editor plugins that are preventing that, if we follow the same approach here https://github.com/godotengine/godot/pull/109361 by having a HFlowContainer with two children LeftHBoxContainer and RightHBoxContainer we can eliminate most of those issues in a clean way, and we will be left with the `EditorTitleBar` preventing that which this PR fixes and it can go with only reusing the new Container for main screen buttons only, and also we can optionally allow disabling the button's text in the EditorSettings like this.

<img width="638" height="47" alt="image" src="https://github.com/user-attachments/assets/cd1f791b-432a-4771-80cb-23b35e2fdc53" />

bruvzg commented:
This pretty much breaks editor title on macOS.

- Window name is no longer visible.
- Large portions of the title are not responding to drag.
- Tabs seems off center.
- Resizing is not quite uniform.

https://github.com/user-attachments/assets/b79e63e9-d802-4795-b629-b20389a11983

Title bar element alignment before the change, window name and all empty areas are draggable:

<img width="1514" height="241" alt="Screenshot 2025-08-07 at 22 48 07" src="https://github.com/user-attachments/assets/50106fe1-b749-44a7-ac9b-8b0472f880f9" />

bruvzg commented:
Some notes:

There's `set_center_control` method in the `EditorTitleBar` used with `main_editor_button_hb`  to force center it regardless of other controls. Current implementation require a direct child of `EditorTitleBar` as argument, so it won't work when it wrapped in the new container, and it also depends on `main_editor_button_hb` being fixed size to work. Forced centering supposed to work only if there's plenty of free space on both sides, so it probably can be adjusted.

Window title can be added to the menu scroll container, since only one of two can be visible at the same time (menu bar still need to be in the scene tree, but will be zero size and invisible).

Window drag can be fixed by setting PASS mouse filter on all new containers.

WhalesState commented:
> Some notes:
> 
> There's `set_center_control` method in the `EditorTitleBar` used with `main_editor_button_hb` to force center it regardless of other controls. Current implementation require a direct child of `EditorTitleBar` as argument, so it won't work when it wrapped in the new container, and it also depends on `main_editor_button_hb` being fixed size to work. Forced centering supposed to work only if there's plenty of free space on both sides, so it probably can be adjusted.
> 
> Window title can be added to the menu scroll container, since only one of two can be visible at the same time (menu bar still need to be in the scene tree, but will be zero size and invisible).
> 
> Window drag can be fixed by setting PASS mouse filter on all new containers.

We may not need this anymore after this PR since it doesn't require the old ways of centering the main screen button and it's managed automatically by only using 3 container and their expanding, the MainScreenButtons now have the higher expand ratio and is still affected by the EditorRunBar and the Renderer Button which are all part of the right hbox now.

bruvzg commented:
It's not always 3 controls, there are extra spacers (left/right_menu_spacer) added on macOS to account for window buttons, these spacers are always different width, so this will introduce misalignment.

I'll retest it tomorrow.

WhalesState commented:
> It's not always 3 controls, there are extra spacers (left/right_menu_spacer) added on macOS to account for window buttons, these spacers are always different width, so this will introduce misalignment.

Removed the right spacer since it's useless now after this change and moved the left spacer into the left HBox container is an internal child since it's only used for MACOS to show the title label, so they are now only 3 children for the EditorTitleBar.

bruvzg commented:
> Removed the right spacer since it's useles

It's not useless, if OS language is right-to-left and Godot language is left-to-right (or vice versa) buttons will be over the right spacer.

WhalesState commented:
> [  X  -  +  ][    left container    ][   center container   ][    right container   ][ ]
>                                            /\                                           
>                                          center

It's so hard to access a mac and it will be great if you can help me by downloading EditorDebugger from asset store then enabling inspecting elements in the plugin and you inspect the title bar with F12 to be able to change the label or container settings in editor inspector to find a quick fix.


The reason why they are not centered is that they are no longer 3 containers, it should be like this to work on mac similar to other platforms.
`[             left container           ] [   center container    ] [           right container          ]`
`[ [  X  -  +  ]  scene-name - Godot    ] [   2D 3D Game Assets   ] [    play buttons   [ right spacer ] ]`

Also for RTL layout i tested it and it works fine in all other platforms.

KoBeWi commented:
You can test it on non-mac too, I think you just have to remove the conditions here:
https://github.com/godotengine/godot/blob/a3b42d85d27668f8992c0779ed3cc82d13db3dd9/editor/editor_node.cpp#L8117-L8127

bruvzg commented:
> You can test it on non-mac too

Not fully, you can enable extra spacers/title, and hide menu bar, but won't get native window buttons, so it would be hard to tell if something is wrong.

> It's so hard to access a mac and it will be great if you can help me by downloading EditorDebugger

Will check it on Monday.

KoBeWi commented:
I did that in https://github.com/godotengine/godot/pull/105106#issuecomment-2784536122 and it was enough to see the problem.

bruvzg commented:
One tiny fix to make title bar drag work:
```diff
diff --git a/editor/editor_node.cpp b/editor/editor_node.cpp
index 78e9ad85bb..84efce00e0 100644
--- a/editor/editor_node.cpp
+++ b/editor/editor_node.cpp
@@ -8193,6 +8193,7 @@ EditorNode::EditorNode() {
                project_title->set_text_overrun_behavior(TextServer::OVERRUN_TRIM_ELLIPSIS);
                project_title->set_mouse_filter(Control::MOUSE_FILTER_PASS);
                project_title_control->add_child(project_title);
+               project_title_control->set_mouse_filter(Control::MOUSE_FILTER_PASS);
                title_left_hbox->add_child(project_title_control, false, INTERNAL_MODE_BACK);
        }
```

The rest looks and works fine on macOS.

WhalesState commented:
> There's still an issue with tiny window sizes, scroll containers seems to work only to some extent:

The 2D/3D/Game/Script/Asset plugins prevents the window from being resized down, I'm using EditorDebugger plugin to hide The HFlowContainer that contains all those buttons for 2D and Game editors to test this PR.

<img width="896" height="84" alt="image" src="https://github.com/user-attachments/assets/d366101d-af61-47c0-a723-90b9480802f0" />

bruvzg commented:
> I'm using EditorDebugger plugin to hide The HFlowContainer that contains all those buttons for 2D and Game editors to test this PR.

Yep, with the part hidden, it seems to resize fine all the way down.

bruvzg commented on editor/gui/editor_scroll_box.cpp:
```diff
@@ -0,0 +1,277 @@
+
+void EditorScrollBox::_notification(int p_what) {
+	switch (p_what) {
+		case NOTIFICATION_TRANSLATION_CHANGED:
```
```suggestion
		case NOTIFICATION_ACCESSIBILITY_UPDATE: {
			RID ae = get_accessibility_element();
			ERR_FAIL_COND(ae.is_null());

			DisplayServer::get_singleton()->accessibility_update_set_role(ae, DisplayServer::AccessibilityRole::ROLE_SCROLL_VIEW);
			DisplayServer::get_singleton()->accessibility_update_add_action(ae, DisplayServer::AccessibilityAction::ACTION_SCROLL_LEFT, callable_mp(this, &EditorScrollBox::_accessibility_action_scroll_left));
			DisplayServer::get_singleton()->accessibility_update_add_action(ae, DisplayServer::AccessibilityAction::ACTION_SCROLL_RIGHT, callable_mp(this, &EditorScrollBox::_accessibility_action_scroll_right));
		} break;
		case NOTIFICATION_TRANSLATION_CHANGED:
```
Mark it as scrollable and add action callbacks.

WhalesState replied:
Will use `ACTION_SCROLL_BACKWARD` and `ACTION_SCROLL_FORWARD` instead of `ACTION_SCROLL_LEFT` and `ACTION_SCROLL_RIGHT` since it also has a vertical layout.

bruvzg replied:
`ACTION_SCROLL_BACKWARD` and `ACTION_SCROLL_FORWARD` were removed in latest version of accesskit, and only kept for compatibility, they are aliases for `UP` and `DOWN`.

bruvzg replied:
So for vertical it should be `if` with LEFT/RIGHT and UP/DOWN branches.

WhalesState replied:
Can we instead use those four action and return early when the action is not supported for vertical or horizontal modes ?

```c++
	void _accessibility_action_scroll_up(const Variant &p_data);
	void _accessibility_action_scroll_down(const Variant &p_data);
	void _accessibility_action_scroll_left(const Variant &p_data);
	void _accessibility_action_scroll_right(const Variant &p_data);
```

bruvzg replied:
I think it should not have all 4 actions registered at the same time, since it will show up as if it has both scroll directions, while there's only one.

WhalesState replied:
> I think it should not have all 4 actions registered at the same time, since it will show up as if it has both scroll directions, while there's only one.

We can register the `EditorScrollBox` as an abstract class and we only add two actions if it's vertical or horizontal.

bruvzg replied:
Yes, it looks fine.

bruvzg commented on editor/gui/editor_scroll_box.cpp:
```diff
@@ -0,0 +1,277 @@
+		p_property.usage = PROPERTY_USAGE_NONE;
+	}
+}
+
```
```suggestion

void EditorScrollBox::_accessibility_action_scroll_left(const Variant &p_data) {
	_scroll(false);
}

void EditorScrollBox::_accessibility_action_scroll_right(const Variant &p_data) {
	_scroll(true);
}

```

bruvzg commented on editor/gui/editor_scroll_box.h:
```diff
@@ -0,0 +1,95 @@
+	ScrollContainer *scroll_container = nullptr;
+	Control *control = nullptr;
+
+	void _scroll(bool p_right);
```
```suggestion
	void _scroll(bool p_right);
	void _accessibility_action_scroll_left(const Variant &p_data);
	void _accessibility_action_scroll_right(const Variant &p_data);
```

KoBeWi commented:
> These buttons are now completely hidden when not running the project, instead of just being disabled. They are only relevant during runtime and do not need to remain visible during regular editing.

That means that when running the project, they will still take space and make the title bar bigger. Hiding them doesn't really improve anything and only makes the play button annoyingly switch places.

WhalesState commented on doc/classes/EditorHScrollBox.xml:
This is mainly for documentation on how to use the new class and make it usable in Editor Plugins. I’ll make a change to address https://github.com/godotengine/godot/issues/109496, and I think it’s worth documenting. If not, we can replace the docs with inline code comments in the class instead.

KoBeWi replied:
Exposing it with EditorPlugins in mind requires a proposal, to clarify use-cases. See https://github.com/godotengine/godot-proposals/issues/3850 for example. Sounds useful, but had pretty much 0 interest, so it was closed. Not being exposed then allowed to [freely change the "API"](https://github.com/godotengine/godot/pull/71685) of ViewPanner.

As for the documentation, the new class looks rather self-explanatory. It's basically a ScrollContainer with extra buttons. We don't usually put doc comments in the code, unless it's a core class with some important considerations.

WhalesState commented:
> That means that when running the project, they will still take space and make the title bar bigger. Hiding them doesn't really improve anything and only makes the play button annoyingly switch places. Unless you never run your project.

I will move this to a new PR to be able to discuss it's usage.

WhalesState commented on doc/classes/EditorScrollBox.xml:
```diff
@@ -0,0 +1,84 @@
+		<method name="get_control" qualifiers="const">
+			<return type="Control" />
+			<description>
+				Returns the control node used by the [ScrollContainer].
```
`Returns the control added to the [ScrollContainer] as a child, otherwise returns [code]null[/code].`

WhalesState commented on editor/editor_node.h:
```diff
@@ -741,6 +741,7 @@ class EditorNode : public Node {
 	static EditorFolding &get_editor_folding() { return singleton->editor_folding; }
 
 	static EditorTitleBar *get_title_bar() { return singleton->title_bar; }
+	static HBoxContainer *get_title_bar_right_hbox() { return singleton->title_right_hbox; }
```
This was added to keep the main screen buttons centered after adding editor plugins to the title bar using `add_control_to_container(EditorPlugin.CONTAINER_TOOLBAR, Control)`.

WhalesState committed:
Extract EditorScrollBox class from BottomPanel and reuse it for the top bar.

Co-authored-by: Pāvels Nadtočajevs <7645683+bruvzg@users.noreply.github.com>
Co-authored-by: A Thousand Ships <96648715+AThousandShips@users.noreply.github.com>
Co-authored-by: Tomasz Chabora <kobewi4e@gmail.com>

WhalesState commented:
> That means that when running the project, they will still take space and make the title bar bigger. Hiding them doesn't really improve anything and only makes the play button annoyingly switch places. Unless you never run your project.

The play bar failed to keep the main screen buttons centered before this change, that's why i have made some contributions to avoid this issue.

Before (This caused parts of the editor to be off-screen):
<img width="1278" height="124" alt="image" src="https://github.com/user-attachments/assets/16e147c8-f21f-4702-95d8-ecffcaf6ba69" />

After (Fixes the off-screen issue caused by the EditorTitleBar but not the main screen buttons centering issue):
<img width="956" height="124" alt="image" src="https://github.com/user-attachments/assets/3cd0cdd9-0a4c-4b27-a546-6dedbffe6898" />


Solutions (Hiding the pause and stop buttons while editing and removing the renderer button from the EditorTitleBar):
<img width="956" height="124" alt="image" src="https://github.com/user-attachments/assets/ab4223a5-f0f2-4704-b0ef-5d99d6d07402" />


This helps to reduce to the right container minimum width to help keep the buttons centered but it doesn't solve the issue:
<img width="1200" height="124" alt="image" src="https://github.com/user-attachments/assets/4f46e9ed-bb07-48ac-95b0-7de7e1e75f17" />

KoBeWi commented on editor/gui/editor_scroll_box.h:
```diff
@@ -0,0 +1,97 @@
+			EditorScrollBox(false) { is_fixed = true; }
+};
+
+class EditorVScrollBox : public EditorScrollBox {
```
Unused class.

KoBeWi commented on editor/gui/editor_scroll_box.cpp:
```diff
@@ -0,0 +1,298 @@
+}
+
+bool EditorScrollBox::_set(const StringName &p_name, const Variant &p_value) {
+	if (p_name == "vertical") {
```
This looks like unused code. The property is not defined.
Also makes `_validate_property()` not called.

WhalesState replied:
> This looks like unused code. The property is not defined. Also makes `_validate_property()` not called.

It extends `BoxContainer` that's why i have this method to update the layout when `BoxContainer::vertical` is changed.

KoBeWi commented on editor/gui/editor_scroll_box.cpp:
```diff
@@ -0,0 +1,298 @@
+		case NOTIFICATION_THEME_CHANGED: {
+			_update_buttons_icon_and_tooltip();
+		}
+			[[fallthrough]];
```
This doesn't make much sense. You can just call `_update_buttons()`, it's a single line either way.

WhalesState commented on editor/gui/editor_scroll_box.cpp:
```diff
@@ -0,0 +1,298 @@
+		enabled_scroll->connect(CoreStringName(changed), update_buttons, CONNECT_DEFERRED);
+	}
+
+	if (!enabled_scroll->is_connected(SceneStringName(value_changed), update_disabled)) {
```
Seems like `BoxContainer` is not meant to be extended, it's better to extend `Container` class instead since there are no signals for orientation changing and `set_vertical` is not a virtual methos.

WhalesState commented on editor/gui/editor_scroll_box.cpp:
```diff
@@ -0,0 +1,199 @@
+	Button *button = get_left_button();
+	button->set_button_icon(theme_cache.arrow_left);
+	button->set_accessibility_name(TTRC("Scroll Left"));
+	button->set_tooltip_text(TTRC("Scroll Left\nHold Ctrl to scroll to the begin.\nHold Shift to scroll one page."));
```
On RTL layout it says scroll to the `begin` instead of `end`, and to fix this it will result in 4 translated lines instead of just 2, so we can consider replacing both translations to say `Hold Ctrl to scroll to the far left` and `Hold Ctrl to scroll to the far right`.

KoBeWi commented on editor/editor_node.cpp:
```diff
@@ -7419,28 +7444,31 @@ void EditorNode::_update_main_menu_type() {
+#endif
 	} else {
+#ifndef MACOS_ENABLED
+		menu_scroll_box = memnew(EditorScrollBox);
```
Main screen plugins _might_ eventually get converted to docks, so the top buttons would become a TabBar.

Repiteo commented:
Superseded by #113051; thanks for your contribution!
